src/tactic/operand.lean

```diff
@@ -0,0 +1,37 @@
+/-
+Copyright (c) 2019 Lucas Allen. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Lucas Allen
+-/
+
+import algebra.big_operators
+import tactic.interactive
+import tactic.converter.interactive
+
+open tactic
+open finset
+
+namespace conv.interactive
+
+meta def operand_core (t : conv.interactive.itactic) : conv unit :=
+do (r, lhs, _) ← target_lhs_rhs,
+   guard (r = `eq),
+   lam ← match lhs with
+     | `(finset.sum  %%S %%lam) := tactic.applyc `finset.sum_congr >> skip >> return lam
```

Comment:
I know very little about meta code and matching expressions etc...
Are these matches up to defeq, or up to synteq?
Otherwise I guess it might be good to include `finsupp.sum` and similar "bigops".

Reply:
@jcommelin The "matching" is being done verbatim against the `expr` representation of the expression.

Reply:
@jcommelin, do you know what else there is besides `finsupp.sum/prod`?

Reply:
Could this work on other `fold`s? Like `list` and `multiset`? (And `foldr`, `foldl`, etc...)

Reply:
For `finsupp.sum` a lemma `finset.sum_congr` is required. I came up with this:

```
import data.finsupp

universes u v w
variables {α : Type u} {β : Type v} {γ : Type w}
[has_zero β] [comm_monoid γ]
variables {s₁ s₂ : finset α} {a : α} {f g : α →₀ β} {op : α → β → γ}


lemma prod_congr (h : f.support = g.support) :
 (∀x∈g.support, f x = g x) → finsupp.prod f op = finsupp.prod g op :=
begin
  library_search
end
```

Unfortunately `library_search` doesn't work. Once this is proved I think it'll be fairly straight forward to extend `operand` to include `finsupp.sum` and it's cousin `finsupp.prod`.

Reply:
I think something like `operand` is unnecessary for `foldl` and `foldr` because these functions fold over the whole list. Whereas `operand` is useful when `f x = g x` for `x \in S`, instead of the stronger `f = g`. I not 100% confident of that, however, maybe it is useful somehow, I dunno.

Reply:
Instead of enumerating all the the functions this could work with (e.g. `finset.sum`), it would be worth considering coming up with an attribute (similar to `extensionality`) to label the lemma that makes this tactic applicable.

Reply:
The problem with labeling lemmas with an attribute is that applying different lemmas gives different sets of goals, and these sets of goals need to be handled individually. In `operand` the cases for `finset.sum` and `finset.prod` call `skip` after applying the lemma `finset.sum_congr` and `finset.prod_congr`, whereas the case for `finset.fold` doesn't have this `skip`.

At this stage I'll just enumerate all the lemmas. However, long term, if the list becomes too large I'd like to revisit this idea, and learn how attributes work.